src/rgw/rgw_dedup.cc

```diff
@@ -0,0 +1,2363 @@
+      // TBD: add support for EC
+      // We need to find the user byte size without the added protection
+      double replica_level = (double)s.num_object_copies / s.num_objects;
+      *p_num_objects       = s.num_objects;
+      *p_num_objects_bytes = s.num_bytes / replica_level;
+      ldpp_dout(dpp, 10) <<__func__ << "::" << pool_name << "::num_objects="
+                         << s.num_objects << "::num_copies=" << s.num_object_copies
+                         << "::num_bytes=" << s.num_bytes << "/" << *p_num_objects_bytes << dendl;
+    }
+    return 0;
+  }
+#if 1
+  //-------------------------------------------------------------------------------
+  //  32B per object-entry in the hashtable
+  //  2MB per shard-buffer
+  //=============||==============||=========||===================================||
+  // Obj Count   || shard count  || memory  ||         calculatio                ||
+  // ------------||--------------||---------||---------------------------------- ||
+  //     1M      ||      4       ||     8MB ||    8MB/32 =  0.25M *   4 =     1M ||
+  //     4M      ||      8       ||    16MB ||   16MB/32 =  0.50M *   8 =     4M ||
+  //-------------------------------------------------------------------------------
+  //    16M      ||     16       ||    32MB ||   32MB/32 =  1.00M *  16 =    16M ||
+  //-------------------------------------------------------------------------------
+  //    64M      ||     32       ||    64MB ||   64MB/32 =  2.00M *  32 =    64M ||
+  //   256M      ||     64       ||   128MB ||  128MB/32 =  4.00M *  64 =   256M ||
+  //  1024M( 1G) ||    128       ||   256MB ||  256MB/32 =  8.00M * 128 =  1024M ||
+  //  4096M( 4G) ||    256       ||   512MB ||  512MB/32 = 16M.00 * 256 =  4096M ||
+  // 16384M(16G) ||    512       ||  1024MB || 1024MB/32 = 32M.00 * 512 = 16384M ||
+  //-------------||--------------||---------||-----------------------------------||
+  static md5_shard_t calc_num_md5_shards(uint64_t obj_count)
+  {
+    // create headroom by allocating space for a 10% bigger system
+    obj_count = obj_count + (obj_count/10);
+
+    uint64_t M = 1024 * 1024;
+    if (obj_count < 1*M) {
+      // less than 1M objects -> use 4 shards (8MB)
+      return 4;
+    }
+    else if (obj_count < 4*M) {
+      // less than 4M objects -> use 8 shards (16MB)
+      return 8;
+    }
+    else if (obj_count < 16*M) {
+      // less than 16M objects -> use 16 shards (32MB)
+      return 16;
+    }
+    else if (obj_count < 64*M) {
+      // less than 64M objects -> use 32 shards (64MB)
+      return 32;
+    }
+    else if (obj_count < 256*M) {
+      // less than 256M objects -> use 64 shards (128MB)
+      return 64;
+    }
+    else if (obj_count < 1024*M) {
+      // less than 1024M objects -> use 128 shards (256MB)
+      return 128;
+    }
+    else if (obj_count < 4*1024*M) {
+      // less than 4096M objects -> use 256 shards (512MB)
+      return 256;
+    }
+    else {
+      return 512;
```

Comment:
So this inherently self-limits sharding and memory allocation?  What happens when there are 10 billion RGW objects?

Reply:
512 shards are enough to cover 16G objects slots in the sharded hash-table.
The dedup skips small objects (currently anything smaller than 4MB) so this limit is not too tight.
Also duplicates are taking a single entry which gives us extra head room
16G objects bigger than 4MB translates to 64PB of user data (not including meta-data and replication)

I can allow for bigger systems, but they will require more memory.
16 billion objects requires 1GB of memory per participating RGW, if we are willing to allocate more memory I can grow the limit (say 2GB for 32 billion objects)

Reply:
>The dedup skips small objects
Ah, good to know.  Is this deduplication at whole-object granularity?

>Also duplicates are taking a single entry which gives us extra head room

IF there are duplicates.  With my operational experience I've learned that any time we think "Nobody would ever actually..." we're deceiving ourselves, in accordance with https://quotefancy.com/quote/1717591/Joel-Garreau-The-Harvard-Law-of-Animal-Behavior-holds-that-under-controlled-experimental

My intent here is ensure that we aren't going to fail when run against very large deployments.